### PR TITLE
Replace mode

### PR DESCRIPTION
**Pull request by rebornix**
Yay! We love PRs! 🎊

Please include a description of your change & check your PR against this list, thanks:

- [x] Commit message has a short title & issue references
- [x] Each commit does a logical chunk of work.  
- [x] It builds and tests pass (e.g `gulp tslint`)

Add a new mode; Replace :)

### CODE REVIEW

File: src/mode/modeHandler.ts [hunk 1]
```diff
@@ -239,6 +242,25 @@ export class SearchState {
...
+  public get replaceCursorStartPosition() {
+    return this._replaceCursorStartPosition;
+  }
+
+  public originalChars: string[] = [];
+
+  constructor(startPosition: Position) {
+    this._replaceCursorStartPosition = startPosition;
+    let text = TextEditor.getLineAt(startPosition).text.substring(startPosition.character);
+    /* tslint:disable:forin */
```
Comment (johnfn) [inline src/mode/modeHandler.ts]: Use for of

**Comment by johnfn** [comment]
Whoa, this is huge! Replace mode has been a big want of mine. Is it ready to go?

**Comment by rebornix** [comment]
@johnfn replace for...in with for...of with returning iterators. I suppose it's ready to go :)

File: src/mode/modeHandler.ts [hunk 2]
```diff
@@ -239,6 +242,24 @@ export class SearchState {
...
+  public get replaceCursorStartPosition() {
+    return this._replaceCursorStartPosition;
+  }
+
+  public originalChars: string[] = [];
+
+  constructor(startPosition: Position) {
+    this._replaceCursorStartPosition = startPosition;
+    let text = TextEditor.getLineAt(startPosition).text.substring(startPosition.character);
+    for (let [key, value] of text.split("").entries()) {
```
Comment (johnfn) [inline src/mode/modeHandler.ts]: That's that ES6 style that I like :-)

**Comment by johnfn** [comment]
Yeah, it generally looks good to me. Two quick questions before merge.

* Can you add some tests?
* Is it repeatable with .? If so, can you test that as well?

**Comment by rebornix** [comment]
@johnfn my bad ... I'll add tests real quick.

**Comment by rebornix** [comment]
Added test cases for Replace Mode and `.` repeat.

The coverage of `.` repeat might be low but it now covers most user cases.

**Comment by rebornix** [comment]
Replace commands can be prefixed with numbers but it's not implemented here, if I implement it before this PR gets merged, I'll update this PR. Otherwise create separate task to track it.

**Comment by johnfn** [comment]
Wonderful work, @rebornix. This is big enough to push out a new version, IMO. I'll give it a spin locally and hopefully get to doing that later tonight. :)

**Pull request merged**